dcc-portal-ui/app/scripts/survivalanalysis/js/service.js

```diff
@@ -0,0 +1,199 @@
+
+        return $q.all({
+          survivalData: fetchSurvival,
+          setsMeta: fetchSetsMeta,
+        })
+          .then(function (responses) {
+            return processResponses(responses);
+          });
+      }
+
+      var defaultHeadingMap = {
+        setName: 'donor_set_name',
+        id: 'donor_id',
+        time: 'time',
+        status: 'donor_vital_status',
+        survivalEstimate: 'survival_estimate',
+      };
+
+      function dataSetToTSV (dataSet, headingMap) {
+        var headings = _({})
+          .defaults(defaultHeadingMap, headingMap)
+          .values()
+          .join('\t');
+
+        _.defaults({}, defaultHeadingMap, headingMap);
+
+        var contents = _(dataSet)
+          .map(function (set) {
+            return set.donors.map(function (donor) {
+              return [set.meta.name, donor.id, donor.time, donor.status, donor.survivalEstimate].join('\t');
+            });
+          })
+          .flatten()
+          .join('\n');
+
+        var tsv = headings + '\n' + contents;
+        return tsv;
+      }
+
+      _.extend(this, {
+        fetchSurvivalData: fetchSurvivalData,
+        dataSetToTSV: dataSetToTSV
+      });
+
+    });
+
+  module.service('SurvivalAnalysisLaunchService', function(SetService, Restangular, LocationService,
+    SetNameService, $timeout, $location, Page){
+      const _service = this;
+
+      var d1, d2, type = 'donor';
+
+      _service.getSetName = function(filters){
+        return SetNameService.getSetFilters()
+          .then(function (filters) {
+            return SetNameService.getSetName(filters);
+          });
+      }
+
+      // Wait for sets to materialize
+      _service.wait = (ids, numTries, callback) => {
+        if (numTries <= 0) {
+          return;
+        }
+        SetService.getMetaData(ids).then(function(data) {
+          var finished = _.filter(data, function(d) {
+            return d.state === 'FINISHED';
+          });
+
+          if (finished.length === ids.length) {
+            callback(data);
+          } else {
+            $timeout(function() {
+              _service.wait(ids, --numTries, callback);
+            }, 1500);
+          }
+        });
+      }
+
+       /* Phenotype comparison only takes in donor set ids */
+      _service.launchAnalysis = (setIds) => {
+        return Restangular
+          .one('analysis')
+          .post('phenotype', setIds, {}, {'Content-Type': 'application/json'})
+          .then(function(data) {
+            if (!data.id) {
+              console.log('Could not retrieve analysis data.id');
+            }
+            LocationService.goToPath('analysis/view/phenotype/' + data.id);
+          }).finally(function(){
+            Page.stopWork();
+          });
+      };
+
+      _service.launchSurvivalAnalysis = (entityType, entityId, entitySymbol, filters) => {
+
+        Page.startWork();
+
+        _service.getSetName(filters).then(function (setName) {
+          d1 = {
```

Comment:
More duplication. We should model the variation and parameterize the generalized version